aten/src/ATen/cpu/vec256/vec256_base.h

```diff
@@ -606,33 +609,65 @@ Vec256<T> inline clamp_min(const Vec256<T> &a, const Vec256<T> &min_vec) {
   return c;
 }
 
-template <class T, typename Op, typename std::enable_if_t<!std::is_integral<T>::value, int> = 0>
+struct Vec256i;
+
+#ifdef __AVX2__
+
+template <class T, typename Op>
 static inline Vec256<T> bitwise_binary_op(const Vec256<T> &a, const Vec256<T> &b, Op op) {
```

Comment:
This only works for integral types, right? I am confused why this modification wasn't put in `vec256_int.h`. Do you know why?

Reply:
I think it should work for any types that are not AVX2-optimized? All it does it to take advantage of int-type avx2 instructions.

Reply:
Err, so I agree you can technically bitwise OR a float, by treating it as some bits and ORing the bits. But I find it hard to believe that anyone would want that. Is that what you mean?

Reply:
That's what I meant. The bitwise operators actually are already available for floating-point numbers on AVX2-enabled CPUs: https://github.com/pytorch/pytorch/blob/2d24005d186844ace9020aa94a5be293f6396f16/aten/src/ATen/cpu/vec256/vec256_float.h#L304-L317

I simply don't see any reason to limit this optimization to non-optimized integer vectors, given the fact that there isn't much difference in the code. There may be potentially creative use of other non-optimized types, such as `half` and `bfloat` (for example changing the sign bit).

Reply:
OK. In that case if you can show me evidence that we aren't doing unnecessary extra loads in the case of integer vectors (e.g., show me the assembly after this change) I'll accept it in this style.

Reply:
OK. Would this answer your question: https://godbolt.org/z/PTdzXU The new implementation seems to lead to much more optimized assemblies.

Reply:
So, first, I agree that the new code is better than what we did before.

But I think it could be even better. Here's why.

We write operations directly on `Vec256` because that makes things compositional. So most of the operations in this class don't do explicit loads/stores to the AVX registers, because we assume that things are already in `__m256`. But your code here violates this assumption (ostensibly, because you are doing the cast), *despite* the fact that you are being passed in `Vec256`.

This seems to matter in practice. Here is a modified version of your Godbolt: https://godbolt.org/z/hycFeD new_bitwise_and3 shows an implementation of bitwise and between three operands. I've shown two versions: one where I implement it using your code (which does the load/store), and another "optimal" version where I write everything by hand. The extra loads and stores show up in the generated assembly.

Here is my claim: if you had put this implementation in the integer VEc256 specialization, this would have removed the necessity for doing the reinterpret cast (which is probably why you were doing the load/store in the first place). Then you could have operated directly on the passed in `__m256i` registers. Then your implementations would have been compositional in the way I describe.

Reply:
It seems to me that we still have some misunderstanding.

1. My main claim is that all integer Vec256 specializations won't be affected by this PR.

> if you had put this implementation in the integer VEc256 specialization, this would have removed the necessity for doing the reinterpret cast (which is probably why you were doing the load/store in the first place).

All bitwise operations on Vec256 that inherits Vec256i (which uses `__m256i` without cast) have been implemented already:

https://github.com/pytorch/pytorch/blob/12fb8148e48d3e39d16832a84ac4b59182a00c4e/aten/src/ATen/cpu/vec256/vec256_int.h#L694-L705

This PR does not delete these implementations and doesn't override any of these. The added functions explicitly excluded situations that Vec256 inherits Vec256i (see the template declaration above the newly defined bitwise operator overloads):

    template<class T, typename std::enable_if_t<!std::is_base_of<Vec256i, Vec256<T>>::value, int> = 0>

Therefore, in case AVX2 is enabled, this PR should only have changed the behavior when there is no specifically optimized AVX2 specialization of `Vec256<T>`, such as `Vec256<int8_t>`, `Vec256<uint8>`, `Vec256<bfloat16>`, etc., until a specifically optimized specialization is implemented.

2. My secondary claim is that the load/store operations won't appear in the assemblies that this PR would generate.

I made a mistake in the previous godbolt link. If the arrays are aligned (as what this PR does), the load and store instructions are gone in the assembly:

https://godbolt.org/z/Jbg7Ke

(See the difference between `new_bitwise_and_noalign` and `new_bitwise_and`.)

Reply:
Ahhh, here we go! OK, this is good enough for me.